MAH FIRST BRANCH EVAH

shmashing opened the pull request.

Commit:
MAH FIRST BRANCH EVAH

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,245 @@
+  draw_board(False)
+
+  master.bind("<Button-1>", left_click_event)
+  master.bind("<Button-2>", flag_square)
```
I noticed in your event handlers that one of the handlers is named after the button used to trigger it (`left_click_event`) and the other is named after what it does (`flag_square`). Can you change them so they are consistent with each other?

Commit:
Round TWO: FIGHT

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+  board.assign_mines()
+
+  game.master.bind("<Button-1>", board.left_click_event)
+  game.master.bind("<Button-2>", board.right_click_event)
```
Out of curiosity, why are these bindings outside the `Game` objects initialization?

shmashing replied:
These are outside because the bindings weren't being recognized in the Game objects initialization. The only way I could get them to work is by having them in the main function.

shmashing replied:
Ah, also because the function they are binding to is a method within the board class and I was getting an error that the global variable board wasn't found

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+
+def main():
+
+  global mines, game, board
```
What does it mean for a variable to be `global` in Python and why are these global?

shmashing replied:
From my understanding - a global variable in python is one that is accessible from every function and class outside of the one you are working in. These are global so that the game class and have access to the board class and vice versa.

jaredonline replied:
Did you consider just putting a reference to `board` within `Game` and vice versa? In general, globals are confusing and bad [=

shmashing replied:
I figured it out. Turns out most of your comments here are related ;) the Game class now owns a Board instance within itself and that is what is being drawn and manipulated. This fixed the bindings issue as well as the global variable issues.

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+      board.__tiles[box_index_x][box_index_y].marked_mine = True
+
+    except IndexError:
+      pass
```
What does the `pass` keyword in Python do?

shmashing replied:
I believe it's just like a skip. In this context, we want to iterate over tiles surrounding a tile, but that runs the risk of throwing an Index Error if we try to look at a tile outside of the board area. So, I'm saying if that happens, then we don't care, just keep on goin!

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+        mines_to_assign -= 1
+
+        for i in range(-1, 2):
+          for j in range(-1, 2):
```
This is hella confusing, but I *think* its grabbing neighboring tiles to increment their `value` field, which is to indicate the number of adjacent mines, right? If so, you might want to write a `Tile#neighbors` method that returns an array of indexes, like: `[[1, 2], [1, 3], [3, 1]]` or whatever and then iterate those.

shmashing replied:
Just curious, but why is that a better solution than just iterating over a box around the tile?

jaredonline replied:
A couple reasons come to mind:
 - wrapping a chunk of confusing code into a method gives it a label and a name which, hopefully, helps illuminate what the code is doing
 - you could actually do some error checking within that method (like only return valid neighbor coordinates)
 - in general there's an idea about how much complexity you can hold in your head at a time. It's hard to reason about this particular method (for me) because this piece of logic pushes me over (my admittedly small) limit of complexity I can manage. If you just get rid of it and replace it with `for neighbor in self.neighbors` I'm like "oh, nice, iterate over the neighbors".

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+
+    mines = NUMBER_MINES - marked_mines
+
+    bottom_text = 'Mines left: ' + repr(mines)
```
What does `repr` do?

shmashing replied:
it's full name is represent. It makes whatever it is called on into what the interpreter is expecting. In this case turning an integer into a string

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+BOARD_WIDTH = 300
+BOX_SIZE = 20
+
+NUMBER_MINES = ((BOARD_WIDTH / BOX_SIZE - 1)**2)//10 
```
What does `//` do?

shmashing replied:
Integer division. This is just to make sure that the number of mines is an integer because we will iterate over it

jaredonline replied:
Do you know how it converts a `Float` to an `Int`?

shmashing replied:
My first thought would be that it just truncates it? So if you have 20.0 for example calling int(20.0) would return 20 or int(20.7) == 20.

jaredonline replied:
You should write a little script and test it.

shmashing replied:
I did. I was right in my assumption

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,256 @@
+
+    self.master.destroy()
+
+    main()
```
Calling `main` from here is *really* confusing. Try to figure a way to not have to do that.

jaredonline reviewed (changes requested):
Some thoughts [=

Commit:
Game class and Tile class changed

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,254 @@
+        if(0 <= (self._x_index + i) < (BOARD_WIDTH/BOX_SIZE - 1)):
+          if(0 <= (self._y_index + j) < (BOARD_WIDTH/BOX_SIZE -1)):
+            #if(i != 0 and j != 0):
+            neighbors.append([self._x_index + i, self._y_index + j])
```
Walk through this loop with a piece of paper and a pencil and write down each step of it, by hand, and you'll see what the issue is.

Commit:
Rearranging the Game Class

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,261 @@
+
+    self._b = Button(self._master, text = 'Play Again?', command = self.new_game)
+    self._panel.create_text(BOARD_WIDTH/2, BOARD_HEIGHT + 40, text  = 'Game Over', fill = 'red') 
+    self._b.pack()
```
So for the stuff above where you're doing `x.a.b()` and you're saying "X shouldn't know about B!"... that's not really true in the `self` context. So these are all okay, IMO. It's like "Game knows it has a thing called `_master`, and it knows how `_master` works". There's some examples I'll comment on below that aren't okay [=

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,261 @@
+    try:
+      self._board._tiles[box_index_x][box_index_y].show_self()
+
+      if(self._board._tiles[box_index_x][box_index_y].value == 0):
```
This is an example that isn't okay. This one is reaching into the private details of `_board` to get information. You can actually tell this by the way you named your variables. `_` is a pattern for languages that don't have the concept of private and public information. If you're calling a method on something that isn't `self` with a leading `_`, you should be wary.

Here, you could say "well... Game needs to know if a tile's value is a bomb... but it doesn't have access directly. What does have access? The board!" So you could write `Board#location_is_bomb(x, y) => bool`, and here, instead of this if, you could write `if (self._board.location_is_bom(box_index_x, box_index_y)`.

Commit:
Attempting to follow Law of Demeter

Commit:
Added: additional features for flagging and unflagging mines
Changed: Game objects no longer directly have access to tile
objects. Game objects now tell board object to change tile
objects.

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,332 @@
+
+  game._board.draw()
+
+  game._master.mainloop()
```
Calling private stuff on objects right off the bat here

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,332 @@
+
+    self._master.destroy()
+
+    self.__init__()
```
Recalling `__init__` on an object is a code smell, you should move the necessary function into something like `initialize_game` and call it instead.

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,332 @@
+    try:
+      self._board.show_tile(box_index_x, box_index_y)
+
+      if(self._board.get_tile_value(box_index_x, box_index_y) == 0):
```
Does python have enums?

shmashing replied:
I'm not exactly sure what an enum is.

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,332 @@
+
+  def check_for_win(self):
+
+    for mine in self._board._mines:
```
Calling a private attribute on an object

shmashing replied:
Okay, so should those objects not be private, then?

jaredonline commented on MineSweeper.py:
```diff
@@ -0,0 +1,332 @@
+  def check_for_win(self):
+
+    for mine in self._board._mines:
+      if(not self._board._tiles[mine[0]][mine[1]].marked_mine):
```
Again, calling a private attribute on an object

Commit:
Version 6... I think...

Added: initialize_game function for the Game objects

Changed: made some variables and objects public so they can
be accessed from other classes.
